File: DSA_python/DSA11029 ĐƯỜNG ĐI TỚI NODE LÁ.py
class TreeNode:
    def __init__(self, parent = 0, child = None):
        self.parent = parent
        self.child = child if child is not None else []

def add_to_tree(edges):
    mp = dict()
    child_set = set()

    for parent, child in edges:
        if parent not in mp: mp[parent] = TreeNode(parent)
        if child not in mp: mp[child] = TreeNode(child)

        mp[parent].child.append(mp[child])
        child_set.add(child)
    
    parents = [parent for parent,_ in edges]
    root = parent

    for parent in parents:
        if parent not in child_set:
            root = parent
    
    return mp[root]

def is_leaves(node):
    if not node: return False
    if len(node.child) == 0: return True
    return False

def traversal(node, ans, now):
    if not node: return

    now.append(node.parent)

    if is_leaves(node) == False:
        for i in node.child:
            traversal(i, ans, now)
    elif is_leaves(node) == True:
        ans.append(now[:])

    now.pop()
    return ans

def cmp(A):
    return A[-1]

def submit():
    n = int(input())
    edges = list()
    for i in range(n-1):
        parent, child = map(int, input().split())
        edges.append((parent, child))

    root = add_to_tree(edges)

    ans = traversal(root, [], [])

    ans.sort(key = cmp)

    for i in ans:
        print(" ".join(map(str, i)))











test = int(input())
for i in range(test):
    submit()

File: DSA/DSA11029 ĐƯỜNG ĐI TỚI NODE LÁ.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    vector<TreeNode*> child;

    TreeNode(int v) : val(v) {}
};

struct point
{
    int parent, child;
};




TreeNode* add(vector<pair<int,int>> &edges)
{
    unordered_map<int,TreeNode*> m;
    unordered_set<int> child_set;
    
    for (auto &&i : edges)
    {
        int parent = i.first, child = i.second;

        if (!m.count(parent)) m[parent] = new TreeNode(parent);
        if (!m.count(child)) m[child] = new TreeNode(child);

        m[parent]->child.push_back(m[child]);
        child_set.insert(child);
    }

    TreeNode* root = nullptr;
    for (auto &&i : m)
    {
        int node_val = i.first;
        if (!child_set.count(node_val))
        {
            root = i.second;
            break;
        }
    }
    
    return root;
}

bool is_leaves(TreeNode *node)
{
    if (!node) return false;
    if (node->child.size() == 0) return true;
    return false;
}

void traversal(TreeNode *node, vector<vector<int>> &ans, vector<int> now)
{
    if (!node)
    {
        return;
    }

    now.push_back(node->val);

    if (!is_leaves(node))
    {
        for (auto &&i : node->child)
        {
            traversal(i, ans, now);
        }
    }
    else if (is_leaves(node))
    {
        ans.push_back(now);
    }
    return;
}

bool cmp(vector<int> &A, vector<int> &B)
{
    return A[A.size()-1] < B[B.size()-1];
}

void submit()
{
    int n;
    cin >> n;

    vector<pair<int,int>> edges(n-1);

    for (int i = 0; i < n-1; i++)
    {
        cin >> edges[i].first >> edges[i].second;
    }

    TreeNode *root = nullptr;
    root = add(edges);
    

    vector<vector<int>> ans;
    vector<int> now;
    traversal(root, ans, now);
    
    sort(ans.begin(), ans.end(), cmp);

    for (auto &&i : ans)
    {
        for (auto &&j : i)
        {
            cout << j << " ";
        }
        cout << endl;
    }
}

int main()
{
    ios::sync_with_stdio(false);                                    // Speed up input/output
    cin.tie(nullptr);
    cout.tie(nullptr);

    int so_bo_test;
    cin >> so_bo_test;
    cin.ignore();
    
    while (so_bo_test--)
    {
        submit();
    }

    return 0;
}